### DISCUSSION TITLE
How to run a custom kernel?

### DISCUSSION THREAD

**Discussion by dimaqq** [lima-vm/lima#1767] [General]
Hi all, I'm new to Lima, and I like it very much so far.

There's one use case I'd like to cover, that is to run newest Linux kernels (6.5.0) or possibly kernels built from tree (e.g. bpf-next).

What template would be most suitable, and how to get the vm to boot the cooked kernel instead of the external kernel?

Also... is Lima a good tool for this job?

**Comment by afbjorklund** [comment]
It might be overkill, you can probably get started using just QEMU (with the -kernel parameter)?

https://qemu-project.gitlab.io/qemu/system/linuxboot.html

But there _was_ a feature added for the "riscv64" platform support, to use a separate `kernel`...

There you can specify a location to be used, instead of whatever is present in the cloud image

**Comment by dimaqq** [comment]
For now, I've started using UTM, booted a livecd, made a "hard disk install" to `/dev/vda` and that seems to work.

I was wondering if I could do this within context of lima, and maybe keep all the other useful integrations (shared dirs, logs, k8s, colima, etc...)

**Reply by afbjorklund** [reply]
The scope depends a bit if you want to replace the kernel of an existing distribution, or if you want to make your distribution (construct your own rootfs as well). The major requirement is that you need support for `cloud-init`, but it can be hacked...

Most distributions allow you to install different kernels using the normal package manager, and then you only need to reboot after installing it. You can either do that in the provision scripts, or make your own image where you replace the kernel.
